Switch to qiskit_algorithms

woodsp-ibm opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Switches from using qiskit.algorithms, whose use is now deprecated, over to qiskit_algorithms.

Resolves #287

### Details and comments

I've listed the items that have been done and tasks under these where changes remain to be done

1. Added qiskit_algorithms to requirements.txt

- [x] Need to update this to specify >= 0.2.0, when that version is released, so the algorithm_globals, that has just been migrated there, can be used and logic changed here from importing it from qiskit.utils to importing it from qiskit_algorithms.utils

2. Added intersphinx to qiskit_algorithms in the docs conf.py, so external refs to that will resolve in the docs

3. Changed imports and any class refs in docstrings from qiskit.algorithms to qiskit_algorithms. This is in code, tutorials and readme.

- [x] Update 01/02 portfolio optimization/diversification tutorials. These cannot be switched at present due to specific type checks in Qiskit Optimization, that they use, which need to be relaxed. They also seed algorithm_globals, which needs changing too - see 4. below.
- [x] Check if the version requirements for qiskit_optimization should be updated to whatever the min version is that it will be that moves over to qiskit_algorithms (0.6.0) (Done via suggestion in comment https://github.com/qiskit-community/qiskit-finance/pull/291#discussion_r1312915917 below)

4. algorithm_globals. I saw some places where its seeded but what is tested does not use any rng from there so I removed these so as to minimize changes that are truly needed in this regard in switching to use the qiskit_algorithms version. Though its a minimal change it seems if we do not need to seed this, as nothing under test draws any random values from it, then we may as well remove the seeding from the test to simplify things. 

- [x] As per 1 above, once qiskit_algorithms 0.2.0 is released and the requirements can be updated, change remaining 

5. There were still a couple of tests using the QuantumInstance so I updated these.

t-imamichi commented on requirements.txt:
```diff
@@ -1,4 +1,5 @@
 qiskit-terra>=0.20.0
+qiskit-algorithms>=0.2.0
 qiskit-optimization>=0.2.0
```
It may require qiskit-optimization>=0.6.0 to support qiskit-algorithms.
```suggestion
qiskit-optimization>=0.6.0
```

woodsp-ibm replied:
Thanks, yes, I had not changed it yet - one of the todos above in the main comment above noted this was yet to be done. I had intended to do this as I transitioned the notebooks over. That would be the first step, the version update, where they should still work with the new version with qiskit.algorithms, as a compatibility check, then switch to qiskit_algorithms to complete things

woodsp-ibm replied:
CI is running here off Opt main branch so pulls in the latest anyway. It all seems to be working so I committed this suggestion first and will make the remaining changes needed for the switch after.

woodsp-ibm commented:
Optimization switched to `qiskit_algorithms` in qiskit-community/qiskit-optimization#537 and this just went through CI fine still using qiskit.algorithms; which was the intent, it working I mean, to accommodate transition of existing code. As Opt now has moved qiskit_algorithms I will update this PR for those aspects I noted and held back on.

woodsp-ibm commented:
I switched the 2 notebooks that used Qiskit Optimization over to using qiskit_algorithms now that Opt has made the change too. I updated the text a little too as well as some of the variable names that were for VQE originally but things were switched over to SamplingVQE when primitive algos were added. It did not change the function just the naming just to remove them simply being called vqe which might, if one wasn't paying attention, cause some confusion - making it sqve for instance I hope allievates that and makes it clearer its SamplingVQE without putting the whole text in which made things a bit verbose - I did for the method in the class though.

woodsp-ibm commented:
I did a couple of tweaks to links and text that had Qiskit's finance that are not really about qiskit_algorithms, but since I had this PR open and I was reviewing the docs CI built from this PR and spotted them I figured I'd just fix them here.

woodsp-ibm commented:
I had removed the on-hold label and the [WIP] from the title as it's now complete/ready,

t-imamichi reviewed (approved):
LGTM
